src/engine/renderer/InternalImage.cpp

```diff
@@ -0,0 +1,1007 @@
+			break;
+
+		case GL_TEXTURE_2D:
+			functionCompressed2D = glCompressedTexImage2D;
+			function2D = glTexImage2D;
+			target = GL_PROXY_TEXTURE_2D;
+			break;
+
+		case GL_TEXTURE_CUBE_MAP:
+			functionCompressed2D = glCompressedTexImage2D;
+			function2D = glTexImage2D;
+			target = GL_PROXY_TEXTURE_CUBE_MAP;
+			break;
+
+		default:
+			ASSERT( false );
+			return 0;
+	}
+
+	if ( texture3D )
+	{
+		if ( textureCompressed )
+		{
+			functionCompressed3D( target, 0, image->internalFormat, image->uploadWidth, image->uploadHeight, 0, border, mipSize, nullptr );
+
+			GL_CheckErrors();
+		}
+		else
+		{
+			function3D( target, 0, image->internalFormat, image->uploadWidth, image->uploadHeight, 0, border, format, type, nullptr );
+
+			GL_CheckErrors();
+		}
+	}
+	else
+	{
+		if ( textureCompressed )
+		{
+			functionCompressed2D( target, 0, image->internalFormat, image->uploadWidth, image->uploadHeight, border, mipSize, nullptr );
+
+			GL_CheckErrors();
+		}
+		else
+		{
+			function2D( target, 0, image->internalFormat, image->uploadWidth, image->uploadHeight, border, format, type, nullptr );
+
+			GL_CheckErrors();
+		}
+	}
+
+	GLint finalWidth;
+	glGetTexLevelParameteriv( target, 0, GL_TEXTURE_WIDTH, &finalWidth );
+
+	if ( finalWidth == 0 )
+	{
+		image->uploadWidth >>= 1;
+		image->uploadHeight >>= 1;
+
+		if ( image->uploadWidth == 1 || image->uploadHeight == 1 )
+		{
+			return 1;
+		}
+
+		return 1 + GetImageHardwareScalingStep( image, format );
+	}
+
+	return 0;
+}
+
+int GetImageCustomScalingStep( image_t *image, imageParams_t *imageParams )
+{
+	int scalingStep = 0;
+
+	// Perform optional picmip operation.
+	if ( !( image->bits & IF_NOPICMIP ) )
+	{
+		int picMin = std::max( -1, r_picMin->integer );
+		int picMax = std::max( 0, r_picMax->integer );
+		int imagePicMin = 0;
+		int imagePicMax = 0;
+
+		if ( imageParams != nullptr )
+		{
+			imagePicMin = std::max( 0, imageParams->picMin );
+			imagePicMax = std::max( 0, imageParams->picMax );
+		}
+
+		// If image has picMax value set, enforce downscaling using this value.
+		if ( imagePicMax != 0 )
+		{
+			picMax = imagePicMax;
+		}
+
+		// If r_picMin == -1 and shader has picMin value set, keep original size.
+		if ( picMin == -1 && imagePicMin != 0 )
+		{
+			picMax = 0;
+		}
+
+		if ( picMax != 0 )
+		{
+			// If image has picMin value set and r_picMin > 0,
+			// use greater value between picMin and picMax.
+			if ( imagePicMin > 0 && picMin > 0 )
+			{
+				picMax = std::max( picMin, picMax );
+			}
+
+			// Do not downscale this image below picMin size.
+			if ( picMax < imagePicMin )
+			{
+				picMax = imagePicMin;
+			}
+
+			// Downscale image to picMax size.
+			int scaledWidth = image->width;
+			int scaledHeight = image->height;
+
+			while ( picMax < scaledWidth || picMax < scaledHeight )
+			{
+				scaledWidth >>= 1;
+				scaledHeight >>= 1;
+				scalingStep++;
+			}
+		}
+
+		scalingStep = std::max( scalingStep, std::max( 0, r_picmip->integer ) );
```

Comment:
I think we better just remove the enforced range of 0 to 3, I see no reason for this enforced range, especially since we may reduce an image more than `r_picmip 3` would do (even, automatically when checking against hardware capabilities). `r_picmip 4` on a 8K texture is still huge. This range, like the whole `r_picmip` feature, assumes we use `128×128` textures everywhere.

Reply:
I added a commit to not enforce that range anymore.